src/java.base/unix/classes/sun/nio/ch/SourceChannelImpl.java

```diff
@@ -107,8 +112,8 @@ protected void implCloseSelectableChannel() throws IOException {
                 assert state == ST_CLOSING;
                 long th = thread;
                 if (th != 0) {
-                    nd.preClose(fd);
                     NativeThread.signal(th);
+                    nd.preClose(fd);
```

Comment:
I think we need a much more detailed description of why the original order is wrong and why it should apply to all "unix" platforms. Further, the same order (preClose then signal) also exists in each of these files
```
src/java.base/share/classes/sun/nio/ch/DatagramChannelImpl.java
src/java.base/share/classes/sun/nio/ch/ServerSocketChannelImpl.java
src/java.base/share/classes/sun/nio/ch/SocketChannelImpl.java
src/java.base/unix/classes/sun/nio/ch/SinkChannelImpl.java
src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpChannelImpl.java
src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java
src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpServerChannelImpl.java
```
Please either correct each of those instances or explain why they don't need similar treatment.

Reply:
Ok, I think you've convinced me that reordering those calls makes sense. Other threads that might retry their read or write operation will fail because the channel has been closed (notice `assert !isOpen();` on line 823.

However, you haven't explained why the same pattern that exists in the classes I listed above doesn't need correcting.

Reply:
@keithc-ca 
The changes are applied to the files you have listed. I have made the changes to those files